File: internal/services/user_service_test.go
package services

import (
	"testing"

	"github.com/hanyouqing/openauth/internal/auth"
	"github.com/hanyouqing/openauth/internal/models"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/assert"
	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
)

func setupTestDBForUser(t *testing.T) *gorm.DB {
	db, err := gorm.Open(sqlite.Open(":memory:"), &gorm.Config{})
	assert.NoError(t, err)

	err = db.AutoMigrate(&models.User{})
	assert.NoError(t, err)

	return db
}

func TestUserService_List(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test users
	for i := 0; i < 5; i++ {
		passwordHash, _ := auth.HashPassword("password123")
		user := models.User{
			Username:     "user" + string(rune(i)),
			Email:        "user" + string(rune(i)) + "@example.com",
			PasswordHash: passwordHash,
			Status:       "active",
		}
		db.Create(&user)
	}

	tests := []struct {
		name     string
		page     int
		pageSize int
		expected int
	}{
		{
			name:     "first page",
			page:     1,
			pageSize: 2,
			expected: 2,
		},
		{
			name:     "second page",
			page:     2,
			pageSize: 2,
			expected: 2,
		},
		{
			name:     "last page",
			page:     3,
			pageSize: 2,
			expected: 1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			users, total, err := service.List(tt.page, tt.pageSize)
			assert.NoError(t, err)
			assert.Equal(t, 5, total)
			assert.LessOrEqual(t, len(users), tt.pageSize)
		})
	}
}

func TestUserService_Get(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test user
	passwordHash, _ := auth.HashPassword("password123")
	user := models.User{
		Username:     "testuser",
		Email:        "test@example.com",
		PasswordHash: passwordHash,
		Status:       "active",
	}
	db.Create(&user)

	tests := []struct {
		name        string
		userID      uint64
		expectError bool
	}{
		{
			name:        "existing user",
			userID:      user.ID,
			expectError: false,
		},
		{
			name:        "non-existing user",
			userID:      999,
			expectError: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := service.Get(tt.userID)
			if tt.expectError {
				assert.Error(t, err)
				assert.Nil(t, result)
			} else {
				assert.NoError(t, err)
				assert.NotNil(t, result)
				assert.Equal(t, tt.userID, result.ID)
			}
		})
	}
}

func TestUserService_Create(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	tests := []struct {
		name        string
		username    string
		email       string
		password    string
		expectError bool
	}{
		{
			name:        "valid user",
			username:    "newuser",
			email:       "newuser@example.com",
			password:    "password123",
			expectError: false,
		},
		{
			name:        "duplicate username",
			username:    "newuser",
			email:       "another@example.com",
			password:    "password123",
			expectError: true,
		},
		{
			name:        "duplicate email",
			username:    "anotheruser",
			email:       "newuser@example.com",
			password:    "password123",
			expectError: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			user, err := service.Create(tt.username, tt.email, tt.password)
			if tt.expectError {
				assert.Error(t, err)
				assert.Nil(t, user)
			} else {
				assert.NoError(t, err)
				assert.NotNil(t, user)
				assert.Equal(t, tt.username, user.Username)
				assert.Equal(t, tt.email, user.Email)
			}
		})
	}
}

func TestUserService_Update(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test user
	passwordHash, _ := auth.HashPassword("password123")
	user := models.User{
		Username:     "testuser",
		Email:        "test@example.com",
		PasswordHash: passwordHash,
		Status:       "active",
	}
	db.Create(&user)

	tests := []struct {
		name        string
		userID      uint64
		data        map[string]interface{}
		expectError bool
	}{
		{
			name:        "update email",
			userID:      user.ID,
			data:        map[string]interface{}{"email": "updated@example.com"},
			expectError: false,
		},
		{
			name:        "update username",
			userID:      user.ID,
			data:        map[string]interface{}{"username": "updateduser"},
			expectError: false,
		},
		{
			name:        "non-existing user",
			userID:      999,
			data:        map[string]interface{}{"email": "test@example.com"},
			expectError: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := service.Update(tt.userID, tt.data)
			if tt.expectError {
				assert.Error(t, err)
				assert.Nil(t, result)
			} else {
				assert.NoError(t, err)
				assert.NotNil(t, result)
			}
		})
	}
}

func TestUserService_Delete(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test user
	passwordHash, _ := auth.HashPassword("password123")
	user := models.User{
		Username:     "testuser",
		Email:        "test@example.com",
		PasswordHash: passwordHash,
		Status:       "active",
	}
	db.Create(&user)

	err := service.Delete(user.ID)
	assert.NoError(t, err)

	// Verify user is deleted
	deletedUser, err := service.Get(user.ID)
	assert.Error(t, err)
	assert.Nil(t, deletedUser)
}

func TestUserService_ChangePassword(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test user
	passwordHash, _ := auth.HashPassword("oldpassword")
	user := models.User{
		Username:     "testuser",
		Email:        "test@example.com",
		PasswordHash: passwordHash,
		Status:       "active",
	}
	db.Create(&user)

	tests := []struct {
		name        string
		userID      uint64
		oldPassword string
		newPassword string
		expectError bool
	}{
		{
			name:        "valid password change",
			userID:      user.ID,
			oldPassword: "oldpassword",
			newPassword: "newpassword123",
			expectError: false,
		},
		{
			name:        "wrong old password",
			userID:      user.ID,
			oldPassword: "wrongpassword",
			newPassword: "newpassword123",
			expectError: true,
		},
		{
			name:        "non-existing user",
			userID:      999,
			oldPassword: "oldpassword",
			newPassword: "newpassword123",
			expectError: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := service.ChangePassword(tt.userID, tt.oldPassword, tt.newPassword)
			if tt.expectError {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}

func TestUserService_UploadAvatar(t *testing.T) {
	db := setupTestDBForUser(t)
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	service := NewUserService(db, logger)

	// Create test user
	passwordHash, _ := auth.HashPassword("password123")
	user := models.User{
		Username:     "testuser",
		Email:        "test@example.com",
		PasswordHash: passwordHash,
		Status:       "active",
	}
	db.Create(&user)

	err := service.UploadAvatar(user.ID, "https://example.com/avatar.jpg")
	assert.NoError(t, err)

	// Verify avatar is updated
	updatedUser, err := service.Get(user.ID)
	assert.NoError(t, err)
	assert.Equal(t, "https://example.com/avatar.jpg", updatedUser.Avatar)
}

File: web/src/services/userService.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { userService } from './userService'
import { apiClient } from './api'

vi.mock('./api')

describe('userService', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('list', () => {
    it('should get user list successfully', async () => {
      const mockResponse = {
        data: [
          { id: 1, username: 'user1', email: 'user1@example.com', status: 'active', created_at: '2025-01-01', updated_at: '2025-01-01' },
          { id: 2, username: 'user2', email: 'user2@example.com', status: 'active', created_at: '2025-01-01', updated_at: '2025-01-01' },
        ],
        total: 2,
        page: 1,
        size: 20,
      }

      vi.mocked(apiClient.get).mockResolvedValue({ data: mockResponse })

      const result = await userService.list(1, 20)

      expect(apiClient.get).toHaveBeenCalledWith('/users', { params: { page: 1, page_size: 20 } })
      expect(result).toEqual(mockResponse)
    })
  })

  describe('get', () => {
    it('should get user by id successfully', async () => {
      const mockResponse = {
        id: 1,
        username: 'testuser',
        email: 'test@example.com',
        status: 'active',
        created_at: '2025-01-01',
        updated_at: '2025-01-01',
      }

      vi.mocked(apiClient.get).mockResolvedValue({ data: mockResponse })

      const result = await userService.get(1)

      expect(apiClient.get).toHaveBeenCalledWith('/users/1')
      expect(result).toEqual(mockResponse)
    })
  })

  describe('create', () => {
    it('should create user successfully', async () => {
      const mockResponse = {
        id: 1,
        username: 'newuser',
        email: 'newuser@example.com',
        status: 'active',
        created_at: '2025-01-01',
        updated_at: '2025-01-01',
      }

      vi.mocked(apiClient.post).mockResolvedValue({ data: mockResponse })

      const result = await userService.create({
        username: 'newuser',
        email: 'newuser@example.com',
        password: 'password123',
      })

      expect(apiClient.post).toHaveBeenCalledWith('/users', {
        username: 'newuser',
        email: 'newuser@example.com',
        password: 'password123',
      })
      expect(result).toEqual(mockResponse)
    })
  })

  describe('update', () => {
    it('should update user successfully', async () => {
      const mockResponse = {
        id: 1,
        username: 'updateduser',
        email: 'updated@example.com',
        status: 'active',
        created_at: '2025-01-01',
        updated_at: '2025-01-01',
      }

      vi.mocked(apiClient.put).mockResolvedValue({ data: mockResponse })

      const result = await userService.update(1, { email: 'updated@example.com' })

      expect(apiClient.put).toHaveBeenCalledWith('/users/1', { email: 'updated@example.com' })
      expect(result).toEqual(mockResponse)
    })
  })

  describe('delete', () => {
    it('should delete user successfully', async () => {
      vi.mocked(apiClient.delete).mockResolvedValue({ data: {} })

      await userService.delete(1)

      expect(apiClient.delete).toHaveBeenCalledWith('/users/1')
    })
  })

  describe('getMe', () => {
    it('should get current user successfully', async () => {
      const mockResponse = {
        id: 1,
        username: 'currentuser',
        email: 'current@example.com',
        status: 'active',
        created_at: '2025-01-01',
        updated_at: '2025-01-01',
      }

      vi.mocked(apiClient.get).mockResolvedValue({ data: mockResponse })

      const result = await userService.getMe()

      expect(apiClient.get).toHaveBeenCalledWith('/users/me')
      expect(result).toEqual(mockResponse)
    })
  })

  describe('updateMe', () => {
    it('should update current user successfully', async () => {
      const mockResponse = {
        id: 1,
        username: 'updateduser',
        email: 'updated@example.com',
        status: 'active',
        created_at: '2025-01-01',
        updated_at: '2025-01-01',
      }

      vi.mocked(apiClient.put).mockResolvedValue({ data: mockResponse })

      const result = await userService.updateMe({ email: 'updated@example.com' })

      expect(apiClient.put).toHaveBeenCalledWith('/users/me', { email: 'updated@example.com' })
      expect(result).toEqual(mockResponse)
    })
  })

  describe('changePassword', () => {
    it('should change password successfully', async () => {
      vi.mocked(apiClient.put).mockResolvedValue({ data: {} })

      await userService.changePassword({
        current_password: 'oldpassword',
        new_password: 'newpassword123',
      })

      expect(apiClient.put).toHaveBeenCalledWith('/users/me/password', {
        current_password: 'oldpassword',
        new_password: 'newpassword123',
      })
    })
  })
})
